modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/TransactionsInMultipleThreadsTest.java

```diff
@@ -0,0 +1,423 @@
+        Assert.assertFalse(cache.containsKey("key2"));
+        Assert.assertFalse(cache.containsKey("key1'"));
+        Assert.assertFalse(cache.containsKey("key2'"));
+
+        cache.removeAll();
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testCrossCacheTransactionInAnotherThread() throws Exception {
+        withAllIsolationsAndConcurrencies(new IgniteClosure<Object, Void>() {
+            @Override public Void apply(Object o) {
+                try {
+                    crossCacheTransactionInAnotherThread();
+                }
+                catch (Exception ignore) {
+                    assert false : "Impossible situation";
+                }
+
+                return null;
+            }
+        });
+    }
+
+    /**
+     * @throws Exception if failed.
+     */
+    private void crossCacheTransactionInAnotherThread() throws Exception {
+        Ignite ignite1 = ignite(0);
+        final IgniteTransactions transactions = ignite1.transactions();
+
+        final IgniteCache<String, Integer> cache = ignite1.getOrCreateCache(cacheConfiguration(null).setName("testCache"));
+        final IgniteCache<String, Integer> cache2 = ignite1.getOrCreateCache(cacheConfiguration(null).setName("testCache2"));
+
+        final Transaction tx = transactions.txStart(transactionConcurrency, transactionIsolation);
+
+        cache.put("key1", 1);
+        cache2.put("key2", 2);
+
+        tx.suspend();
+
+        IgniteInternalFuture<Boolean> fut = GridTestUtils.runAsync(new Callable<Boolean>() {
+            @Override public Boolean call() throws Exception {
+                Assert.assertNull(transactions.tx());
+                Assert.assertEquals(TransactionState.SUSPENDED, tx.state());
+
+                tx.resume();
+
+                Assert.assertEquals(TransactionState.ACTIVE, tx.state());
+
+                cache.put("key3", 3);
+                cache2.remove("key2");
+
+                tx.commit();
+
+                return true;
+            }
+        });
+
+        try {
+            fut.get();
+        }
+        catch (IgniteCheckedException e) {
+            e.printStackTrace();
+            fail("Transaction in another thread failed");
+        }
+
+        Assert.assertEquals(TransactionState.COMMITTED, tx.state());
+        Assert.assertEquals((long)1, (long)cache.get("key1"));
+        Assert.assertEquals((long)3, (long)cache.get("key3"));
+        Assert.assertFalse(cache2.containsKey("key2"));
+
+        cache2.removeAll();
+        cache.removeAll();
+    }
+
+    /**
+     * @throws Exception  if failed.
+     */
+    public void testCrossCacheTransactionInAnotherThreadContinued() throws Exception {
+        withAllIsolationsAndConcurrencies(new IgniteClosure<Object, Void>() {
+            @Override public Void apply(Object o) {
+                try {
+                    crossCacheTransactionInAnotherThreadContinued();
+                }
+                catch (Exception ignore) {
```

Comment:
Let's rename "ignore" to "ignired"
Let's replace "// igniore." with "// No-op."